src/program.js

```diff
@@ -429,6 +429,15 @@ Example: $0 --help run.
       default: true,
       type: 'boolean',
     },
+    'filename': {
+      alias: 'n',
+      describe: 'Name of the created extension package file.',
+      default: undefined,
+      normalize: true,
+      demandOption: false,
+      requiresArg: false,
```

Comment:
#### requiresArg: true

This should be `requiresArg: true`, otherwise the user would be allowed to call it without any argument which doesn't make sense (it defaults to the default template, but still not something we need to support).

#### Explicitly disallow multiple --filename option in the same command

Unfortunately yargs would still allow to call `-n` more than once and the parsed filename property would become an array.
In the current version of this patch, that is going to make `getPackageNameFromTemplate` to throw because `filenameTemplate` will be an Array:
```
TypeError: filenameTemplate.replace is not a function
    at getPackageNameFromTemplate (/home/rpl/mozlab/projects/webext-tools/web-ext/dist/webpack:/src/cmd/build.js:132:40)
    at defaultPackageCreator (/home/rpl/mozlab/projects/webext-tools/web-ext/dist/webpack:/src/cmd/build.js:205:5)
    at build (/home/rpl/mozlab/projects/webext-tools/web-ext/dist/webpack:/src/cmd/build.js:296:18)
```

to be fair there are other options that we should explicitly check to be sure that they are not called more than once, but we could take care of just this one in this pull request (and we will take care of the others in some follow ups).
Unfortunately it doesn't seem that yargs provide a configuration option specifically for this, but we could use the `coerce` option to do it in a way that can be then easily reused (in follow ups) for the other options.
How about something like this:
```js

function throwUsageErrorIfArray(errorMessage: string): any {
  return (value: any): any => {
    if (Array.isArray(value)) {
      throw new UsageError(errorMessage);
    }
    return value;
  };
}

...

    'filename': {
      alias: 'n',
      describe: 'Name of the created extension package file.',
      default: undefined,
      demandOption: false,
      requiresArg: true,
      type: 'string',
      coerce: throwUsageErrorIfArray(
        'Multiple --filename/-n option are not allowed'
      ),
    },
  });
....
```

The unfortunate part of this change is that I guess that it would make coveralls unhappy ;-)
To "make coveralls happy again", we could add an integration test for this as part of `tests/functional/test.cli.build.js`.

Reply:
> The unfortunate part of this change is that I guess that it would make coveralls unhappy ;-)
To "make coveralls happy again", we could add an integration test for this as part of tests/functional/test.cli.build.js.

Done, but `test-functional` doesn't seem to be run during `npm run test-coverage` ? Maybe if I push to my branch Github will pick up the test-functional run as well?

Reply:
Fixed, by exporting the function and adding coverage in the `test.program.js` unit test.

Reply:
@escapewindow
My bad, I forgot that we are not collecting the coverage from test-functional, and I recall that to make it do so there were some issues that I didn't have the time to look into at the time (and now I don't even recall what it was in enough detail).

Looking to the coveralls report:

https://coveralls.io/builds/30660260/source?filename=src%2Fprogram.js#L348

it seems that it is only complaining because on the new `throwUsageErrorIfArray`, in particular about the `if (Array.isArray(...)) {...}` branch.

Let's just export the helper function and test that in a unit test added to `tests/unit/test.program.js`.
That should be a small and simple test and enough to make coveralls to stop complaining.

My apologies again for forgetting that the functional test would have not be enough to make coveralls happy with the coverage (but we can keep the test, yargs has the bad habit to be very surprising in subtle changes from version to version, I prefer if we can have a proper integration test for this).

About normalize, it is totally ok to remove it, anyway we did already determine that it was not going to do anything needed or useful.

Reply:
Yeah, agreed. Looks like we came to the same solution :)

I think this is ready for re-re-review. Thanks!